src/Microsoft.IdentityModel.Protocols.OpenIdConnect/OpenIdConnectProtocolValidator.cs

```diff
@@ -557,37 +668,39 @@ protected virtual void ValidateNonce(OpenIdConnectProtocolValidationContext vali
         /// </summary>
         /// <param name="validationContext">a <see cref="OpenIdConnectProtocolValidationContext"/> that contains the 'state' to validate.</param>
         /// <exception cref="ArgumentNullException">if 'validationContext' is null.</exception>
+        /// <exception cref="OpenIdConnectProtocolInvalidStateException">if 'state' is present in <see cref="OpenIdConnectProtocolValidationContext.State"/> but either <see cref="OpenIdConnectProtocolValidationContext.ProtocolMessage"/> or <see cref="OpenIdConnectProtocolValidationContext.ProtocolMessage.State"/> is null.</exception>
         /// <exception cref="OpenIdConnectProtocolInvalidStateException">if 'state' in the context does not match the state in the message.</exception>
         protected virtual void ValidateState(OpenIdConnectProtocolValidationContext validationContext)
         {
-
             if (validationContext == null)
             {
                 LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10000, GetType() + ": validationContext"), typeof(ArgumentNullException), EventLevel.Verbose);
             }
 
-            // if state is missing, but not required just return. Otherwise process it.
-            if (string.IsNullOrEmpty(validationContext.State))
+            if (validationContext.ProtocolMessage == null)
             {
-                if (RequireState)
-                {
-                    LogHelper.Throw(string.Format(CultureInfo.InvariantCulture, LogMessages.IDX10332, GetType() + ": validationContext"), typeof(OpenIdConnectProtocolInvalidStateException), EventLevel.Error);
-                }
-                else
-                {
-                    return;
-                }
+                LogHelper.Throw(LogMessages.IDX10333, typeof(OpenIdConnectProtocolException), EventLevel.Error);
             }
 
-            // 'state' was sent, but message does not contain 'state'
-            if (validationContext.ProtocolMessage == null || string.IsNullOrEmpty(validationContext.ProtocolMessage.State))
+            // if state is missing, but not required just return. Otherwise process it.
+            if (!RequireStateValidation && string.IsNullOrEmpty(validationContext.State) && string.IsNullOrEmpty(validationContext.ProtocolMessage.State))
```

Comment:
Shouldn't this just be `if (!RequireStateValidation) { return; }`? Otherwise it requires message.State to be absent even if validation is disabled.